### PR TITLE
feat: Add markdown-it-plugin

### PR DESCRIPTION
**Pull request by GordonSmith**
## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised new issues to address them separately

### CODE REVIEW

**Comment by GordonSmith** [comment]
FYI - This code is what drives the embedding of interactive visualizations within VitePress (and other) generated sites.  See https://gordonsmith.github.io/hpcc-js-docs/ for the POC

File: packages/markdown-it-plugins/esbuild.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { nodeTpl, browserTpl } from "@hpcc-js/esbuild-plugins";
+
+//  config  ---
+await Promise.all([
+    nodeTpl("src/index.ts", "dist/index.node", { externalx: ["@observablehq/runtime", "@observablehq/stdlib", "@observablehq/inspector"] }),
+    browserTpl("src/index.ts", "dist/index", { externalx: ["@observablehq/runtime", "@observablehq/stdlib", "@observablehq/inspector"] })
```
Comment (jeclrsg) [inline packages/markdown-it-plugins/esbuild.js:6]: externalx?

File: packages/markdown-it-plugins/src/fence.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import type MarkdownIt from "markdown-it";
+import type { Options } from "markdown-it";
+import type Token from "markdown-it/lib/token.mjs";
+import type Renderer from "markdown-it/lib/renderer.mjs";
```
Comment (jeclrsg) [inline packages/markdown-it-plugins/src/fence.ts:4]: Seems weird that these can't all be one include, but that's just how this library has their stuff structured?

File: packages/markdown-it-plugins/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        return retVal;
+    };
+}
+
+function hookVitepressRender(md: MarkdownIt) {
+
+    const originalRender = md.render;
+    md.render = (src: string, env?: any): string => {
+        const myEnv = env ?? {};
+        let retVal = originalRender.call(md, src, myEnv);
```
Comment (jeclrsg) [inline packages/markdown-it-plugins/src/index.ts:22]: why's this different from how hookRender does the env stuff above?

File: packages/markdown-it-plugins/src/util.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    return fetchInfo.exec == true && fetchInfo.hide !== true;
+}
+
+let idx = 0;
+function calcPlaceholders(content: string, fenceInfo: FenceInfo): RenderNode[] {
+    const retVal: RenderNode[] = [];
+    ++idx;
+    try {
+        const cellNb = ojs2notebook(content);
+        for (let i = 0; i < cellNb.nodes.length; ++i) {
```
Comment (jeclrsg) [inline packages/markdown-it-plugins/src/util.ts:39]: this looks like it always steps over the first node, is that correct?

File: packages/markdown-it-plugins/tests/index.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+const minimalMd = fetch("/tests/Basic usage/Minimal Example.md").then(response => {
+    return response.text();
+}).catch(e => {
+    console.error("Error: ", e.message ?? e);
+});
+const minimalMdText = await minimalMd;
+
+const minimalHtml = md.render("" + await minimalMd);
+// expect(minimalHtml).to.be.a.string;
```
Comment (jeclrsg) [inline packages/markdown-it-plugins/tests/index.ts:22]: some commented out code in here, looked like it was moved into the index.browser.spec.ts

**Review by jeclrsg — COMMENTED** [review]
@GordonSmith just a few minor comments. only thing that seemed like it might be a bug was that for loop index in util.ts?

File: packages/markdown-it-plugins/esbuild.js (hunk 1 above)
Comment (GordonSmith) [reply packages/markdown-it-plugins/esbuild.js:6]: Yea should be removed.

File: packages/markdown-it-plugins/src/fence.ts (hunk 2 above)
Comment (GordonSmith) [reply packages/markdown-it-plugins/src/fence.ts:4]: I agree...  But at least they are only types (we don't bundle marked-it in our dist folder)

File: packages/markdown-it-plugins/src/index.ts (hunk 3 above)
Comment (GordonSmith) [reply packages/markdown-it-plugins/src/index.ts:22]: Up to that point both are the same (and I could re-structure to make cleaner), its how the Observable render gets triggered that is different depending on the host environment.

File: packages/markdown-it-plugins/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        if (env[ENV_KEY]?.length) {
+            const content = encodeURI(JSON.stringify(env[ENV_KEY]));
+            retVal += `<NotebookComponent content="${content}" />`;
```
Comment (GordonSmith) [inline packages/markdown-it-plugins/src/index.ts:25]: This is specific to the VitePress implementation, on reflection I might want to include the "NotebookComponent" in this packages as a resource...

File: packages/markdown-it-plugins/src/util.ts (hunk 4 above)
Comment (GordonSmith) [reply packages/markdown-it-plugins/src/util.ts:39]: It shouldn't, both `for (let i = 0; i < cellNb.nodes.length; ++i)` and `for (let i = 0; i < cellNb.nodes.length; i++)` are the same thing, FWIW in old c++ compilers the ++i was was more efficient, so its just a habit.

(assuming that is what your picking up on)?

File: packages/markdown-it-plugins/src/util.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            const id = `fence-${idx}-${i + 1}`;
+            const content = cellNb.nodes[i].value;
+            retVal.push({
+                ...fenceInfo,
+                id,
+                content,
+                innerHTML: `\
+        <span id="${id}" >
+            ${renderExecutedSrc(fenceInfo) ? content : ""}
+        </span>`
```
Comment (GordonSmith) [inline packages/markdown-it-plugins/src/util.ts:49]: I should remove the leading spaces here.

File: packages/markdown-it-plugins/tests/index.ts (hunk 5 above)
Comment (GordonSmith) [reply packages/markdown-it-plugins/tests/index.ts:22]: Yea I will remove comment code, it was actually copied from the spec file to here...

**Review by jeclrsg — APPROVED** [review]

**Pull request merged**